lib/scholar/linear/isotonic_regression.ex

```diff
@@ -0,0 +1,419 @@
+
+    {x_output, y_output, sample_weights_output, index}
+  end
+
+  defnp contiguous_isotonic_regression(y, sample_weights, max_size) do
+    y_size = max_size
+    target = Nx.iota({Nx.axis_size(y, 0)}, type: :s64)
+    type_wy = Nx.Type.merge(Nx.type(y), Nx.type(sample_weights))
+
+    {{y, target}, _} =
+      while {{y, target},
+             {i = 0, sample_weights, sum_w = Nx.tensor(0, type: Nx.type(sample_weights)),
+              sum_wy = Nx.tensor(0, type: type_wy), prev_y = Nx.tensor(0, type: type_wy), _k = 0,
+              terminating_flag = 0, y_size}},
+            i < y_size + 1 and not terminating_flag do
+        k = target[i] + 1
+
+        cond do
+          k == y_size + 1 ->
+            {{y, target}, {i, sample_weights, sum_w, sum_wy, prev_y, k, 1, y_size}}
+
+          y[i] < y[k] ->
+            i = k
+
+            {{y, target}, {i, sample_weights, sum_w, sum_wy, prev_y, k, terminating_flag, y_size}}
+
+          true ->
+            sum_wy = sample_weights[i] * y[i]
+            sum_w = sample_weights[i]
+
+            {y, sample_weights, i, target, sum_w, sum_wy, _prev_y = prev_y, k,
+             _inner_terminating_flag,
+             y_size} =
+              while {y, sample_weights, i, target, sum_w, sum_wy, prev_y, k,
+                     inner_terminating_flag = 0, y_size},
+                    not inner_terminating_flag do
+                prev_y = y[k]
+                sum_wy = sum_wy + sample_weights[k] * y[k]
+                sum_w = sum_w + sample_weights[k]
+                k = target[k] + 1
+
+                {y, sample_weights, target, i, inner_terminating_flag} =
+                  if k == y_size + 1 or prev_y < y[k] do
+                    y = Nx.indexed_put(y, Nx.new_axis(i, 0), sum_wy / sum_w)
+                    sample_weights = Nx.indexed_put(sample_weights, Nx.new_axis(i, 0), sum_w)
+                    target = Nx.indexed_put(target, Nx.new_axis(i, 0), k - 1)
+                    target = Nx.indexed_put(target, Nx.new_axis(k - 1, 0), i)
+
+                    i =
+                      if i > 0 do
+                        target[i - 1]
+                      else
+                        i
+                      end
+
+                    {y, sample_weights, target, i, 1}
+                  else
+                    {y, sample_weights, target, i, 0}
+                  end
+
+                {y, sample_weights, i, target, sum_w, sum_wy, prev_y, k, inner_terminating_flag,
+                 y_size}
+              end
+
+            {{y, target}, {i, sample_weights, sum_w, sum_wy, prev_y, k, terminating_flag, y_size}}
+        end
+      end
+
+    {y, _} =
+      while {y, {target, i = 0, _k = 0, max_size}}, i < max_size + 1 do
+        k = target[i] + 1
+        indices = Nx.iota({Nx.axis_size(y, 0)})
+        in_range? = Nx.logical_and(i + 1 <= indices, indices < k)
+        y = Nx.select(in_range?, y[i], y)
+        i = k
+        {y, {target, i, k, max_size}}
+      end
```

Comment:
The big question is if any of these loops can be vectorized. Thoughts? :)

Reply:
Actually, I thought about it. But the main problem is that in each iteration, we use data from the previous iteration which is problematic to vectorize. @polvalente @jonatanklosko, maybe you would have some thoughts?

Reply:
Nothing stands out to me, the iterations seem to be sequentially dependent because of the chain lookup in `target`.

Reply:
But target isn't updated, so maybe it's possible

Reply:
Target is updated in every loop of inner while which implies that is updated on every loop of outer while as well

Reply:
It's not updated in this specific loop though :) @polvalente The thing is that for second iteration we have `i = target[0] + 1`, for third one we have `target[target[0] + 1] + 1`, for forth `target[target[target[0] + 1] + 1] + 1` and so on. I don't see how we could compute `i` vector in a single step.

Reply:
Ah, that makes more sense. I also don't see how it could be vectorized

Reply:
Oh, ok, I forgot that we have 3 while loops :sweat_smile: